crates/ruff_db/src/system/memory_fs.rs

```diff
@@ -237,6 +237,34 @@ impl MemoryFileSystem {
         let normalized = SystemPath::absolute(path, &self.inner.cwd);
         normalized.into_utf8_path_buf()
     }
+
+    pub(crate) fn read_directory(
+        &self,
+        path: impl AsRef<SystemPath>,
+    ) -> Result<impl Iterator<Item = Result<DirectoryEntry>> + '_> {
```

Comment:
I see the value in doing that generally, but here I wonder if an opaque return type might actually be better? If you're e.g. relying on the exact type being returned from this method in a test (for example), you might get into difficulties if you later switch the test to using the `OsSystem` and find that a different type is returned from that struct's `read_directory()` method. I think there's some value in saying that the only API guarantee we give here is that some kind of iterator is returned.

Reply:
If the argument is API compatibility with `System` than the method should also return a `Box<dyn...>` (and accept `&SystemPath` instead of `AsRef<SystemPath>` as the argument). I think it's rare that we would switch between systems in tests and the change would be minimal. But having a concrete type can be helpful for a system implementation that's based on the memory file system (e.g. WASM).

Anyway, I don't feel strongly (except that we shouldn't change the return type to `Box<dyn>`)

Reply:
I don't feel strongly either. But I propose we add it when we need it